GitHub Actions CI Port

illusion0001 opened the pull request:
Cache is supported on the penguin
Cold: https://github.com/illusion0001/rpcs3/actions/runs/10871089443/attempts/1
Warm: https://github.com/illusion0001/rpcs3/actions/runs/10871089443

Plans for later, build in clang and gcc on Windows and OSX and releases to upload repo

Margen67 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,143 @@
+        with:
+          name: RPCS3 for Windows
+          path: build/*.7z
+          compression-level: 0
```
`compression-level: 0` doesn't stop it from double-zipping, but it should at least make uploading artifacts faster.
```suggestion
          if-no-files-found: error
          compression-level: 0
```

illusion0001 replied:
I guess I could just glob all the files at least for windows so there's no need to double zip it

illusion0001 replied:
Added zip artifact for this

hcorion commented:
What's the motivation for porting to GitHub Actions? Doesn't it have the exact same restrictions as Azure?

illusion0001 commented:
> What's the motivation for porting to GitHub Actions? Doesn't it have the exact same restrictions as Azure?

so users don't have to sign up to extra services to run auto builds

q4a commented:
> What's the motivation for porting to GitHub Actions? Doesn't it have the exact same restrictions as Azure?

I think that GitHub Actions is useful now because of free arm64 runners. I improved illusion0001's version a bit:
https://github.com/q4a/rpcs3/commits/arm-ci/
But don't understand how `$COMPILER` env to docker?
Check fail on github actions:
https://github.com/q4a/rpcs3/actions/runs/12852962562/job/35835578225
vs sucess on Cirrus CI:
https://cirrus-ci.com/task/6366890357948416

illusion0001 commented:
`$COMPILER` is used to distinguish between GCC and clang builds

q4a commented:
> But don't understand how to pass `$COMPILER` env to docker?

I know that)
Cirrus CI has this lines in log:
https://cirrus-ci.com/task/6366890357948416
```
+ export CC=clang-18
+ export CXX=clang++-18
```
Actions x86_64 has this lines in log:
https://github.com/q4a/rpcs3/actions/runs/12852962562/job/35835578464
```
+ export CC=clang-18
+ export CXX=clang++-18
```
But Actions arm64 has only this lines in log:
https://github.com/q4a/rpcs3/actions/runs/12852962562/job/35835578225
```
+ export CC=
+ export CXX=
```

q4a commented:
Ok. I fixed arm64 build with simple `chmod +x ‎.ci/build-linux-aarch64.sh`.
@illusion0001 can you cherry-pick or just copy my changes to this PR?
https://github.com/q4a/rpcs3/commit/914fd4f9ee2c04a4206a6d3d8ec28a3430d94e99
Here is build result:
https://github.com/q4a/rpcs3/actions/runs/12853858021

Darkhost1999 commented:
![Screenshot.jpg](https://github.com/user-attachments/assets/ed3ca08c-7338-4c49-b37f-628a2de7b7b1)

Maybe instead of matrix: Linux Build it can be matrix: Linux Arch 64 and matrix: Linux X64 by creating new name categories.
I was looking into splitting the files and then just using `uses ./.workflows/linux_aarch64.yml`
That way there's a main file for rpcs3.yml and then maintainable smaller jobs. Especially after adding MacOS then Windows Arm.

illusion0001 commented:
> Ok. I fixed arm64 build with simple `chmod +x ‎.ci/build-linux-aarch64.sh`. @illusion0001 can you cherry-pick or just copy my changes to this PR? https://github.com/q4a/rpcs3/commits/arm-ci/ Here is build result: https://github.com/q4a/rpcs3/actions/runs/12853858021

Pushed

illusion0001 commented:
Arm gcc build is clang, needs fixing

illusion0001 commented:
> > Arm gcc build is clang, needs fixing https://github.com/illusion0001/rpcs3/actions/runs/12855079318/job/35840192563
> 
> I tested few changes and this one works pretty good: [q4a@3e2d9f7](https://github.com/q4a/rpcs3/commit/3e2d9f757b199a242805bba4a5493b8e4714c1ba) CI result: https://github.com/q4a/rpcs3/actions/runs/12875043123

Updated. Also fixed windows artifact uploading build files

Jack54guythecoder commented:
> What's the motivation for porting to GitHub Actions? Doesn't it have the exact same restrictions as Azure?

You can use nightly.link with it.

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+      - name: Append Compiler version to appimage
+        run: |
+          file=".ci/deploy-linux.sh"
+          sed -i 's/RPCS3_APPIMAGE="rpcs3-v${COMM_TAG}-${COMM_COUNT}-${COMM_HASH}_linux64.AppImage"/RPCS3_APPIMAGE="rpcs3-v${COMM_TAG}-${COMM_COUNT}-${COMM_HASH}-${COMPILER}_linux64.AppImage"/g' $file
```
It's to separate the artifacts from being overwritten, i.e clang finishes before gcc, when GCC is uploaded, clang appimage is overwritten. I didn't want to modify script to keep compatibility for other ci

illusion0001 replied:
Well, now that I had a second look at this, it's not needed after all because artifact have compiler name in it

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+        shell: bash
+        run: .ci/setup-windows.sh
+
+      - name: Export Variables
```
It can be done in bash, as for making it a file, I'm afraid that `github.env` won't be resolved there, hence why it's a step.

hcorion replied:
Ahh that's unfortunate, if you could rewrite in bash that'd be great for consistency.

illusion0001 replied:
Updated this step to bash

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+        run: |
+          mkdir -p build
+          .ci/deploy-windows.sh
+          mkdir -p bin2/${{ env.BUILD_RAW }}
```
Bin2 is really only for zipped version artifact to not be double compressed. I'll remove it in favor of 7z

illusion0001 replied:
Zip removed

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+          mkdir -p bin2/${{ env.BUILD_RAW }}
+          mv bin/* bin2/${{ env.BUILD_RAW }}
+
+      - name: Upload artifacts (zip)
```
I'll remove zip

illusion0001 replied:
Zip removed

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+            ${{ matrix.cfg.build_sh }}
+
+      - name: Upload artifacts
+        uses: actions/upload-artifact@main
```
Added
```yaml
        if: |
          github.event_name != 'pull_request' &&
          github.repository == 'RPCS3/rpcs3' &&
          github.ref == 'refs/heads/master' &&
          matrix.compiler == 'clang'
```

illusion0001 replied:
> I'm not familiar with `actions/upload-artifact@main`, is this going to upload binaries to RPCS3/rpcs3-binaries-win/linux/linux-arm64/etc ? I don't see where we set any variables related to the upload repository, so I don't think this will work how we want it to.

Nope, there will have to be a separate step for it. Using appropriate tokens like azure to upload it

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+      - name: Pack up build artifacts
+        shell: bash
+        run: |
+          mkdir -p build
```
Should be sorted

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+          path: bin2
+          if-no-files-found: error
+
+      - name: Upload artifacts (7z)
```
Should be sorted

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+      matrix:
+        compiler: [clang, gcc]
+        cfg:
+          - os: ubuntu-24.04-arm
```
Should be fixed

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+        uses: actions/cache@main
+        with:
+          path: ${{ env.CACHE_DIR }}
+          key: ${{ runner.os }}-${{ env.COMPILER }}-${{ env.QT_VER }}-${{ env.VULKAN_SDK_SHA }}-llvm.lock-glslang.lock
```
Will check

hcorion replied:
Let me know what the outcome is here.

illusion0001 commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,167 @@
+      DEPLOY_APPIMAGE: true
+      APPDIR: "/rpcs3/build/appdir"
+      ARTDIR: "/root/artifacts"
+      RELEASE_MESSAGE: "/rpcs3/GitHubReleaseMessage-${{ matrix.compiler }}.txt"
```
Removed

hcorion commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,171 @@
+      - name: Checkout repository
+        uses: actions/checkout@main
+        with:
+          submodules: recursive
```
We don't want to checkout any submodules here since we do filtered checkout of submodules in the build script.

illusion0001 replied:
Removed

hcorion commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,171 @@
+      - name: Checkout repository
+        uses: actions/checkout@main
+        with:
+          submodules: recursive
```
Same here, we do a filtered submodule checkout in the setup script.

illusion0001 replied:
Removed

Megamouse commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,165 @@
+      matrix:
+        include:
+          - os: ubuntu-24.04
+            docker_img: "rpcs3/rpcs3-ci-focal:1.9"
```
```suggestion
            docker_img: "rpcs3/rpcs3-ci-jammy:1.0"
```

Megamouse commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,165 @@
+            build_sh: "/rpcs3/.ci/build-linux.sh"
+            compiler: clang
+          - os: ubuntu-24.04
+            docker_img: "rpcs3/rpcs3-ci-focal:1.9"
```
```suggestion
            docker_img: "rpcs3/rpcs3-ci-jammy:1.0"
```

Megamouse commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,165 @@
+            build_sh: "/rpcs3/.ci/build-linux.sh"
+            compiler: gcc
+          - os: ubuntu-24.04-arm
+            docker_img: "rpcs3/rpcs3-ci-focal-aarch64:1.0"
```
```suggestion
            docker_img: "rpcs3/rpcs3-ci-jammy-aarch64:1.0"
```

hcorion commented on .github/workflows/rpcs3.yml:
```diff
@@ -0,0 +1,165 @@
+        uses: actions/cache@main
+        with:
+          path: ${{ env.CACHE_DIR }}
+          key: ${{ runner.os }}-${{ env.COMPILER }}-${{ env.QT_VER }}-${{ env.VULKAN_SDK_SHA }}-llvm.lock-glslang.lock
```
Give this a try
```suggestion
          key: ${{ runner.os }}-${{ env.COMPILER }}-${{ env.QT_VER }}-${{ env.VULKAN_SDK_SHA }}-${{ hashFiles(llvm.lock) }}-${{ hashFiles(glslang.lock) }}
```

illusion0001 replied:
Seems to be working after small fixes 👍

illusion0001 commented:
Rebased
